index.html

```diff
@@ -854,23 +862,49 @@ <h3>Extensions to the <code>Performance</code> Interface</h3>
 '!PERFORMANCE-TIMELINE-2/#dfn-performance-entry-buffer'>performance
 entry buffer</a>.</li>
 <li>Increase <a>resource timing buffer current size</a> by 1.</li>
+<li>Return.</li>
 </ol>
 </li>
-<li>If <a>resource timing buffer current size</a> is greater than
-or equal to <a>resource timing buffer size limit</a> and the
-<a>resource timing buffer full flag</a> is false, run the following
+<li>If <a>resource timing buffer full flag</a> is false, run the following
 substeps:
-<ol style='list-style-type: lower-latin;' data-link-for=
-"Performance">
-<li>Set the <a>resource timing buffer full flag</a> to true.</li>
+<ol style="list-style-type: lower-latin;">
+<li>Set <a>resource timing buffer full flag</a> to true.</li>
+<li>Queue a task to run <a>fire a buffer full event</a>.</li>
+</ol>
+<li>If <a>resource timing secondary buffer current size</a> is less than
+<a>resource timing buffer size limit</a>, add <i>new entry</i> to the
```

Comment:
We don't check condition in our implementation since the script wouldn't have a chance to clear the buffer until it receives the event. The secondary buffer is basically unbounded for us.

Reply:
The point of the buffer size limit is to limit the amount of memory used. Having an unbounded secondary buffer makes that limit pointless.

Reply:
Well, if we're delaying the event dispatching to the next task, how are author scripts supposed to know to clear the buffer? This unboundedness only occurs until the next task, and I don't see any reason why it would be bad. The script authors can just allocate a bunch of arrays to do the same if any.

Reply:
While I agree that the unboundedness here would be short lived, I prefer to avoid scenarios where the event is called again on every microtask. So I prefer to define the buffer as bounded to the same size as the RT buffer to guaranty that if the developer either cleared the buffer or doubled its size, all entries will be successfully copied.

This can result in dropped entries when the buffer is too small, but I doubt that would happen in practice, and it makes the processing model simpler.

Reply:
> Well, if we're delaying the event dispatching to the next task, how are author scripts supposed to know to clear the buffer?

I'm not sure I understand the issue.
The proposed model is:
* An entry which overflows the buffer, triggers the event queueing.
* The entry as well as following ones go into the secondary buffer.
* Once the event is fired, developers should make room in the primary buffer (either clear or bump its size)
* Right after that, entries are copied from secondary to primary if possible

Do you see a problem with that model?

Reply:
@rniwa Just removing this check would cause the second buffer to be filled up indefinitely in the case when ResourceTiming is not used at all. We would need a different check.

@yoavweiss I think the problem being addressed is that potentially the secondary buffer could fill up before the bufferfull event is fired. This would cause the developer to lose some entries with nothing they can do about it.

Reply:
I agree that losing entries is a theoretical risk here. I doubt this is something that will happen in real life, assuming developers haven't shrunk their buffer sizes to be unreasonably small.

Reply:
The problem is that the event is dispatched asynchronously. If some script or the HTML/CSS/JS parser ends up requesting more than 150 requests at once and they're cached or failed instantly, then the script may never have a chance to see those entries.

It's not true that this secondary buffer would indefinitely grow when resource timing is never used. If the script never reacts to the event, then we would drop all entries in the secondary buffer after dispatching the very first event.

Reply:
I'm looking at this PR, not at WebKit's implementation. If you want an unbounded secondary buffer then you cannot just clear this buffer after firing a buffer full event. So yes, it would grow indefinitely.

If we go with an unbounded secondary buffer then I would suggest clearing it only if resource timing buffer current size is equal to resource timing buffer size limit, right after firing the resourcetimingbufferfull event (because this is a strong hint that the entries are not being used).

Reply:
You're just picking on my words at that point. Clearly, you have to take my feedback as a whole. What we're proposing is to keep adding entries to the secondary buffer (we should really call it the backup buffer to be consistent with the backup queue we have for custom elements), and then the first time we fire an event on it, if the primary buffer size didn't shrink or the limit hadn't increased, then we would drop all remaining entries in the backup buffer. There won't be no indefinite growth of the backup buffer beyond the first event firing at that point.

Reply:
It wasn't clear to me how you proposed getting rid of secondary buffer entries when they were unused. Maybe I missed something? But I think that what you said just now matches what I proposed (looking at the primary buffer size right after firing the event). SGTM

Reply:
So, if we have a scenario where `X` entries are accumulated every Task, and on every call to `resourcetimingbufferfull` the developer is increasing the buffer by `Y`, where `Y < X`, this scheme will end up consuming an infinite amount of memory.

That seems like a fairly large risk. The trade-off here, where if the developer didn't dedicate enough space for the entries, they get lost, seems better.

The only problematic scenario is if the initial buffer is significantly smaller then the rate of incoming entries per task. That doesn't seem like a likely scenario with the current browser defaults.

Reply:
Oh now I remember why non-first `resourcetimingbufferfull` events are fired synchronously. This is so that we can ensure that either the buffer is emptied after a single task, or that the remaining entries are dropped on the floor.

Not that if we end up enqueuing more `ResourceTiming` entries as a result of dispatching `resourcetimingbufferfull`, we can fall into an infinite loop here but that's no worse than the current status quo, and this is no different from someone causing a mutual recursion between event listeners.

Reply:
What is the benefit of adding all that complexity and edge cases? Why is it all of the sudden OK to fire the events synchronously the second time, but not the first time?

I think a simpler solution would be to basically tell people:
* Don't set the buffer size to ridiculously small amounts. I doubt many people *decrease* the buffer sizes from their default size.
* Clear or significantly increase (at least double) the buffer when it's full

I can see an argument in favor of adding console messages when this happens, so that developers can catch such scenarios, if they at all exist. But adding a ton of complexity in order to save a few entries in theory seem excessive.

Reply:
> Why is it all of the sudden OK to fire the events synchronously the second time, but not the first time?

Because we're already in a new task. There is no security or implementation issues from firing an event synchronously.

Reply:
> Because we're already in a new task

new task relative to what?


> There is no security or implementation issues from firing an event synchronously.

Can you expand on why you objected to the origin sync event firing that was in the spec? What were the issues there?

Reply:
The problem is this. There is a lot of code in the browser engine where the resource load gets canceled (e.g. because the document is getting removed from a frame), and adding a resource timing entry in those timing is not permissible.

However, once this new timer fires, there is nothing like that happening down in the stack frame. So it's totally okay to synchronously fire events.